infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java

```diff
@@ -0,0 +1,60 @@
+package org.corfudb.infrastructure.logreplication.infrastructure;
+
+import org.corfudb.infrastructure.logreplication.infrastructure.plugins.ILogReplicationVersionAdapter;
+import org.corfudb.runtime.collections.TxnContext;
+
+/**
+ * Rolling upgrade handling means cluster must function in a mode where not all
+ * nodes are running the same codebase.
+ * Here the newer code base must function in a backward compatible mode until all
+ * the nodes have been upgraded to the same version.
+ * Once upgrades are done it may atomically migrate data, clean up old code and switch
+ * completely to newer logic.
+ *
+ * To facilitate this in LR we add this module which is intended to be used as follows
+ * on LR startup and when upgrade is complete.
+ *
+ * For anything post corfu-0.4.0.1 we consider the code base to be at "V2"
+ * Practically this means any code that is running Active/Standby logic is considered "V1"
+ * Any code shipped that uses Source/Sink can be thought of as "V2"
```

Comment:
Does this mean the workflow is this handler is mainly for the upgrade from "product with pre corfu-0.4.0.1" to "product with post corfu-0.4.0.1"? I think it is for generic rolling upgrade from "version1" to "version2" right?

Reply:
yes. it is for upgrade from `corfu-0.4.0.1` to the current branch. Note that right now the current branch is master but in the future we could branch off into something like `corfu-0.4.1.0` that's what we are referring to V2 here.